### PR TITLE
Hometask4: Testing

### PR DESCRIPTION
**Pull request by ketanthakkar**
- Created test using enzyme and jest.
- Used snapshot testing
- Used Cypress
- Used Coverage

### CODE REVIEW

File: coverage/clover.xml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (thetreew) [inline coverage/clover.xml]: Coverage information should not be stored in repository - please add to gitignore.

File: cypress.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (thetreew) [inline cypress.json]: Please remove if this file is not used.

File: cypress/fixtures/example.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "Using fixtures to represent data",
+  "email": "hello@cypress.io",
+  "body": "Fixtures are a great way to mock data for responses to routes"
```
Comment (thetreew) [inline cypress/fixtures/example.json]: Same here - please remove if not used.

File: cypress/integration/examples/hometask.e2e.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+describe("Cypress e2e testing", () => {
+    it('should open page', () => {
```
Comment (thetreew) [inline cypress/integration/examples/hometask.e2e.js]: Please update test description - this test checks page title.

File: cypress/integration/examples/hometask.e2e.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+        cy.visit('http://localhost:4000/');
+
+        cy.title().should('include', "ReactJS App");
+    });
+
+    it('should has value', () => {
```
Comment (thetreew) [inline cypress/integration/examples/hometask.e2e.js]: Please update test description with information about the value (what value is this?)

File: package.json [hunk 6]
```diff
@@ -9,7 +9,14 @@
     "open:src": "webpack-dev-server --hot --port 4000",
     "build:dev": "webpack --env dev --progress --profile --colors",
     "build:prod": "webpack --env prod --progress --profile --colors",
-    "test": "jest --watchAll"
+    "test": "jest --watchAll",
+    "test:coverage": "npm run test -- --coverage",
+    "cypress:open": "cypress open"
```
Comment (thetreew) [inline package.json]: What do you think about renaming `cypress:open` to `test:e2e`?
In order to keep test commands grouped.

File: src/components/Content.js [hunk 7]
```diff
@@ -2,10 +2,7 @@ import React from 'react';
 import Filter from './Filter';
 import MovieItem from './MovieItem';
 
-const Content = (props) => {
-    const movies = props.movies.data;
-    const records = props.movies.limit;
-    
+const Content = ({ movies, records }) => {
```
Comment (thetreew) [inline src/components/Content.js]: 👍

File: test/Content.test.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+describe('should render Content component', () => {
+    const tree = shallow(<Content movies={movieData} records={10} />);
+    test('Snapshot test', () => {
+        expect(tree).toMatchSnapshot();
+    });
+});
+
+test('should render Content component with records', () => {
+    const tree = mount(<Content movies={movieData} records={10} />);
+    expect(tree.prop('records')).toEqual(10);
```
Comment (thetreew) [inline test/Content.test.js]: I'd suggest changing this test to check, if value provided in `records` prop is properly rendered by component.

Checking, if prop value is equal to provided value, tests if React works properly, not component.

File: test/ErrorBoundary.test.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      expect(tree).toMatchSnapshot()
+    })
+  
+    it('displays error message on error generated by child', () => {
+      const wrapper = shallow(
+        <ErrorBoundary > 
+          <ComponentWithError />
+        </ErrorBoundary> 
+        )
+      console.log(wrapper.debug() )
```
Comment (thetreew) [inline test/ErrorBoundary.test.js]: Missing `expect` statement.

File: test/ErrorBoundary.test.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+      return (
+        <div>
+          <input type = "text" value = {null}/>  
+        </div>
+      );
+    }
+}
+
+describe('<ErrorBoundary> window',()=> {
+    it('should match the snapshot', () => {
```
Comment (thetreew) [inline test/ErrorBoundary.test.js]: I'd suggest choosing one (`it` or `test`) and sticking to it everywhere (consistency).

File: test/ErrorBoundary.test.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+      const tree = renderer.create(<ErrorBoundary>Test</ErrorBoundary> ).toJSON()
```
Comment (thetreew) [inline test/ErrorBoundary.test.js]: There are different methods used to render component (`renderer.create` and `shallow/mount`).
I'd suggest sticking to one, if possible (to be consistent).

File: test/Main.test.js [hunk 12]
```diff
@@ -0,0 +1,3 @@
+test('', () => {
+    expect("");
```
Comment (thetreew) [inline test/Main.test.js]: Please remove useless tests.

File: test/Movie.test.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import { mount } from 'enzyme';
+import Movie from "../src/components/Movie";
+
+describe('should renders Movie correctly', () => {
```
Comment (thetreew) [inline test/Movie.test.js]: `renders` -> `render`

File: test/Filter.test.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+import { mount } from 'enzyme';
+import Filter from "../src/components/Filter";
+
+describe('should renders Filter correctly', () => {
+  const tree = mount(<Filter movieCount={10}/>);
```
Comment (thetreew) [inline test/Filter.test.js]: Try using `shallow` (instead of `mount`) whenever possible - it will help you testing on unit level and is a bit faster.

**Comment by thetreew** [comment]
General comment: use `describe` function to group related tests - https://jest-bot.github.io/jest/docs/api.html#describename-fn

In some cases `describe` is used as `test`

**Comment by ketanthakkar** [comment]
@thetreew Changes done as per you comment. :) And thanks for suggestion.

File: cypress/integration/examples/hometask.e2e.js [hunk 15]
```diff
@@ -1,11 +1,11 @@
 describe("Cypress e2e testing", () => {
-    it('should open page', () => {
+    it('should test app', () => {
         cy.visit('http://localhost:4000/');
 
         cy.title().should('include', "ReactJS App");
     });
 
-    it('should has value', () => {
+    it('elements should has some values', () => {
```
Comment (thetreew) [inline cypress/integration/examples/hometask.e2e.js]: Let me recommend one technique to verify the quality of test descriptions - values in all `describe`s and `it`s should reveal application's features.

```
Cypress e2e testing
  should test app
  elements should has some values
```
Considering above, how much do we know about the application?

Comment (thetreew) [reply cypress/integration/examples/hometask.e2e.js]: Same technique can be applied to unit tests - "test report" should reveal units' (components') capabilities.

File: test/Filter.test.js [hunk 16]
```diff
@@ -1,15 +1,17 @@
...
   });
-});
 
-describe('should render Filter component with movieCount', () => {
-    const tree = mount(<Filter movieCount={10} />);
-    expect(tree.prop('movieCount')).toEqual(10);
+  test('should render component with movieCount', () => {
+    const count = 10;
+    const tree = shallow(<Filter movieCount={count}/>);
+    expect(tree.find('.results').text()).toEqual(`${count} movies found`);
```
Comment (thetreew) [inline test/Filter.test.js]: Very good 👍 This is what should be tested here.

One hint: try not to replicate component's logic in tests and keep assertions simple.
```js
expect(tree.find('.results').text()).toEqual('10 movies found');
```

Comment (ketanthakkar) [reply test/Filter.test.js]: Done

File: cypress/integration/examples/hometask.e2e.js (hunk 15 above)
Comment (ketanthakkar) [reply cypress/integration/examples/hometask.e2e.js]: > Let me recommend one technique to verify the quality of test descriptions - values in all `describe`s and `it`s should reveal application's features.
> 
> ```
> Cypress e2e testing
>   should test app
>   elements should has some values
> ```
> Considering above, how much do we know about the application?

Done

File: package.json [hunk 17]
```diff
@@ -9,7 +9,14 @@
     "open:src": "webpack-dev-server --hot --port 4000",
     "build:dev": "webpack --env dev --progress --profile --colors",
     "build:prod": "webpack --env prod --progress --profile --colors",
-    "test": "jest --watchAll"
+    "test": "jest --watchAll",
+    "test:coverage": "npm run test -- --coverage",
```
Comment (thetreew) [inline package.json]: I'd recommend adding one command to run tests without watch mode.
```
"test": "jest", // to run tests once
"test:watch": "npm run test -- --watchAll", // to run tests in watch mode
"test:coverage": "npm run test -- --coverage", // to run tests once and generate coverage
```

**Comment by thetreew** [comment]
> @thetreew Done changes.

Looks good to me! 👍

**Pull request merged**